flink-runtime/src/main/java/org/apache/flink/runtime/util/ScheduledFutures.java

```diff
@@ -0,0 +1,58 @@
+package org.apache.flink.runtime.util;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import org.apache.flink.shaded.guava18.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Utils related to {@link ScheduledFuture scheduled futures}.
+ */
+public class ScheduledFutures {
+
+	private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder()
+		.setNameFormat("Flink-Scheduled-Future-SafeGuard")
+		.setDaemon(true)
+		.build();
+
+	/**
+	 * Guard {@link ScheduledFuture} with scheduled future uncaughtException handler, because
+	 * {@link java.util.concurrent.ScheduledExecutorService} does not respect the one assigned to
+	 * executing {@link Thread} instance.
+	 *
+	 * @param scheduledFuture Scheduled future to guard.
+	 * @param uncaughtExceptionHandler Handler to call in case of uncaught exception.
+	 * @param <T> Type the future returns.
+	 * @return Future with handler.
+	 */
+	public static <T> ScheduledFuture<T> withUncaughtExceptionHandler(
+		ScheduledFuture<T> scheduledFuture,
+		Thread.UncaughtExceptionHandler uncaughtExceptionHandler) {
+		final Thread safeguardThread = THREAD_FACTORY.newThread(() -> {
+			try {
+				scheduledFuture.get();
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
```

Comment:
Is it necessary to catch this here? We do not catch it below, so we can also not catch it here.

Reply:
We need to catch it as the `Runnable` does not have `InterruptedException in its signature`, we can replace `Thread.currentThread().interrupt();` with `// noop` though.

Reply:
That makes sense, but I don't think we can replace `Thread.currentThread().interrupt();` with a noop, we still want to pass on the interrupt.